File: C++/207 Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int n = numCourses;

        vector<vector<int>> g(n, vector<int>());
        for(auto &v : prerequisites) {
            auto a = v.front(), b = v.back();
            g[b].push_back(a);
        }

        vector<int> colors(n, 0);
        auto dfs = [&](this auto &&dfs, int u) -> bool {
            colors[u] = 1;
            for(auto &v : g[u]) {
                if(colors[v] == 1 || colors[v] == 0 && dfs(v)) return true;
            }
            colors[u] = 2;
            return false;
        };

        for(int i = 0; i < n; i++) {
            if(colors[i] == 0 && dfs(i)) return false;
        }
        return true;
    }
};

File: Swift/207 Course Schedule.swift
class Solution {
    func canFinish(_ numCourses: Int, _ prerequisites: [[Int]]) -> Bool {
        let n = numCourses
        var g: [[Int]] = .init(repeating: [], count: n)
        for (idx, val) in prerequisites.enumerated() {
            if let a = val.first, let b = val.last {
                g[b].append(a)
            }
        }

        var colors: [Int] = .init(repeating: 0, count: n)
        func dfs(_ u: Int) -> Bool {
            colors[u] = 1
            for item in g[u] {
                if colors[item] == 1 || colors[item] == 0 && dfs(item) {
                    return true
                }
            }
            colors[u] = 2
            return false
        }

        for item in 0 ..< n {
            if colors[item] == 0 && dfs(item) {
                return false
            }
        }

        return true
    }
}
